tfhe/src/core_crypto/algorithms/lwe_multi_bit_programmable_bootstrapping.rs

```diff
@@ -336,228 +324,67 @@ pub fn multi_bit_blind_rotate_assign<Scalar, InputCont, OutputCont, KeyCont>(
-
-        let mut buffers = ComputationBuffers::new();
-
-        buffers.resize(
-            add_external_product_assign_scratch::<Scalar>(
-                multi_bit_bsk.glwe_size(),
-                multi_bit_bsk.polynomial_size(),
-                fft,
-            )
-            .unwrap()
-            .unaligned_bytes_required(),
-        );
-
-        let mut src_idx = 1usize;
-
-        for _ in 0..multi_bit_bsk.multi_bit_input_lwe_dimension().0 {
-            src_idx ^= 1;
-            let idx = rx.recv().unwrap();
-            let (ready_lock, condvar, multi_bit_fourier_ggsw) =
-                &fourier_multi_bit_ggsw_buffers[idx];
-
-            let (src_ct, mut dst_ct) = if src_idx == 0 {
-                (ct0.as_view(), ct1.as_mut_view())
-            } else {
-                (ct1.as_view(), ct0.as_mut_view())
-            };
-
-            dst_ct.as_mut().fill(Scalar::ZERO);
-
-            let mut ready = ready_lock.lock().unwrap();
-            assert!(*ready);
-
-            let multi_bit_fourier_ggsw = multi_bit_fourier_ggsw.lock().unwrap();
-            add_external_product_assign(
-                dst_ct,
-                multi_bit_fourier_ggsw.as_view(),
-                src_ct,
-                fft,
-                buffers.stack(),
-            );
-            drop(multi_bit_fourier_ggsw);
-
-            *ready = false;
-            // Wake a single producer thread sleeping on the condvar (only one will get to work
-            // anyways)
-            condvar.notify_one();
-        }
-
-        if src_idx == 0 {
-            ct0.as_mut().copy_from_slice(ct1.as_ref());
-        }
-
-        let ciphertext_modulus = ct0.ciphertext_modulus();
-        if !ciphertext_modulus.is_native_modulus() {
-            // When we convert back from the fourier domain, integer values will contain up to 53
-            // MSBs with information. In our representation of power of 2 moduli < native modulus we
-            // fill the MSBs and leave the LSBs empty, this usage of the signed decomposer allows to
-            // round while keeping the data in the MSBs
-            let signed_decomposer = SignedDecomposer::new(
-                DecompositionBaseLog(ciphertext_modulus.get_custom_modulus().ilog2() as usize),
-                DecompositionLevelCount(1),
-            );
-            ct0.as_mut()
-                .iter_mut()
-                .for_each(|x| *x = signed_decomposer.closest_representable(*x));
+        Self {
+            switched_modulus_input_lwe_body,
+            switched_modulus_input_mask_per_group,
         }
-
-        threads.into_iter().for_each(|t| t.join().unwrap());
-    });
+    }
 }
 
-/// Deterministic version of [`multi_bit_blind_rotate_assign`].
-pub fn multi_bit_deterministic_blind_rotate_assign<Scalar, InputCont, OutputCont, KeyCont>(
-    input: &LweCiphertext<InputCont>,
+pub fn multi_bit_blind_rotate_assign<Scalar, OutputCont, KeyCont>(
+    switched_modulus_input: &MultiBitmodulusSwitchedCt,
```

Comment:
I don't want to change the meaning of the blind rotate

For the modulus switch decompression, the problem lies in the specialization of a blind rotation to manage some "error correcting terms" which I think could be done with something looking like the multi bit PBS modulus switch zipped with the error correction factors or by "just" updating the prepare multi big GGSW by taking an option for error correction, that would be hard coded to None everywhere we don't do the decompression, as it would be hard coded to None the compiler will be able to delete all the code that does not correspond to that case outside of the decompression.

The factorization makes it  much harder to reason about the multi bit PBS now with the callbacks for the determinism

I would prefer keeping the deterministic and non deterministic variants separate as it makes quite a bit of state to remember inside that function

The AtomicUsize is good to keep

Reply:
I could probably remove commit `refactor(core): factorize multi_bit blind_rotate` without changing much else and keeping commit `refactor(core): separate multi_bit modulus_switch and blind_rotate`
What do you think about this?

Reply:
could be I haven't checked the specific commits while doing the review, I'll take another look